File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/diagnosiskeydist/apispec/TwoHourlyZIPSubmissionPeriod.java
package uk.nhs.nhsx.diagnosiskeydist.apispec;

import uk.nhs.nhsx.diagnosiskeydist.agspec.ENIntervalNumber;

import java.time.Duration;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.*;

import static java.lang.String.format;
import static java.time.ZoneOffset.UTC;
import static java.time.temporal.ChronoField.*;
import static java.time.temporal.ChronoUnit.HOURS;
import static uk.nhs.nhsx.core.Preconditions.checkState;

public class TwoHourlyZIPSubmissionPeriod implements ZIPSubmissionPeriod {
    private static final String TWO_HOURLY_PATH_PREFIX = "distribution/two-hourly/";
    private static final int TOTAL_TWO_HOURLY_ZIPS = ENIntervalNumber.MAX_DIAGNOSIS_KEY_AGE_DAYS * 12;
    private static final DateTimeFormatter HOURLY_FORMAT = DateTimeFormatter.ofPattern("yyyyMMddHH").withZone(UTC);

    private final Instant periodEndDateExclusive;

    public TwoHourlyZIPSubmissionPeriod(Instant dailyPeriodEndDate) {
        periodEndDateExclusive = requireValid(dailyPeriodEndDate);
    }

    private static Instant requireValid(Instant endDate) {
        var date = endDate.atZone(UTC);
        checkState(date.get(HOUR_OF_DAY) % 2 == 0);
        checkState(date.get(MINUTE_OF_HOUR) == 0);
        checkState(date.get(SECOND_OF_MINUTE) == 0);
        checkState(date.get(NANO_OF_SECOND) == 0);
        return endDate;
    }

    /**
     * @return end date (exclusive) of the two-hourly period comprising the Diagnosis Keys posted to the Submission Service at <code>diagnosisKeySubmission</code>
     */
    public static TwoHourlyZIPSubmissionPeriod periodForSubmissionDate(Instant diagnosisKeySubmission) {
        var hour = diagnosisKeySubmission
            .truncatedTo(HOURS)
            .atZone(UTC)
            .getHour();

        return new TwoHourlyZIPSubmissionPeriod(
            diagnosisKeySubmission
                .truncatedTo(HOURS)
                .atZone(UTC)
                .withHour(hour - (hour % 2))
                .plus(Duration.ofHours(2))
                .toInstant()
        );
    }

    @Override
    public String zipPath() {
        return TWO_HOURLY_PATH_PREFIX + twoHourlyKey() + ".zip";
    }

    private String twoHourlyKey() {
        return HOURLY_FORMAT.format(periodEndDateExclusive);
    }

    /**
     * @return true, if <code>diagnosisKeySubmissionDate</code> is covered by the two-hourly period represented by (<code>twoHourlyDate</code> shifted by <code>twoHourlyDateOffsetMinutes</code>)
     */
    @Override
    public boolean isCoveringSubmissionDate(Instant diagnosisKeySubmission, Duration offset) {
        var toExclusive = periodEndDateExclusive
            .plus(offset);

        var fromInclusive = toExclusive
            .minus(Duration.ofHours(2));

        return (diagnosisKeySubmission.isAfter(fromInclusive) || diagnosisKeySubmission.equals(fromInclusive)) && diagnosisKeySubmission.isBefore(toExclusive);
    }

    /**
     * @return list of valid <code>TwoHourlyPeriod</code> ending with <code>this</code> <code>TwoHourlyPeriod</code>
     */
    @Override
    public List<TwoHourlyZIPSubmissionPeriod> allPeriodsToGenerate() {
        var periods = new ArrayList<TwoHourlyZIPSubmissionPeriod>();
        var currentInstant = periodEndDateExclusive;
        for (var i = 0; i < TOTAL_TWO_HOURLY_ZIPS; i++) {
            periods.add(new TwoHourlyZIPSubmissionPeriod(currentInstant));
            currentInstant = currentInstant.minus(Duration.ofHours(2));
        }
        return periods;
    }

    @Override
    public Instant getStartInclusive() {
        return periodEndDateExclusive.minus(Duration.ofHours(2));
    }

    @Override
    public Instant getEndExclusive() {
        return periodEndDateExclusive;
    }

    @Override
    public String toString() {
        return format("2 hours: from %s (inclusive) to %s (exclusive)",
            HOURLY_FORMAT.format(getStartInclusive()),
            HOURLY_FORMAT.format(getEndExclusive()));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TwoHourlyZIPSubmissionPeriod that = (TwoHourlyZIPSubmissionPeriod) o;
        return Objects.equals(periodEndDateExclusive, that.periodEndDateExclusive);
    }

    @Override
    public int hashCode() {
        return Objects.hash(periodEndDateExclusive);
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/diagnosiskeydist/apispec/TwoHourlyZIPSubmissionPeriod.kt
package uk.nhs.nhsx.diagnosiskeydist.apispec

import uk.nhs.nhsx.diagnosiskeydist.agspec.ENIntervalNumber
import java.time.Duration
import java.time.Instant
import java.time.ZoneOffset.UTC
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoField.HOUR_OF_DAY
import java.time.temporal.ChronoField.MINUTE_OF_HOUR
import java.time.temporal.ChronoField.NANO_OF_SECOND
import java.time.temporal.ChronoField.SECOND_OF_MINUTE
import java.time.temporal.ChronoUnit.HOURS
import java.util.*

data class TwoHourlyZIPSubmissionPeriod(val periodEndDateExclusive: Instant) : ZIPSubmissionPeriod {

    init {
        requireValid(periodEndDateExclusive)
    }

    override fun zipPath(): String = "$TWO_HOURLY_PATH_PREFIX${twoHourlyKey()}.zip"

    private fun twoHourlyKey(): String = HOURLY_FORMAT.format(periodEndDateExclusive)

    /**
     * @return true, if `diagnosisKeySubmissionDate` is covered by the two-hourly period represented by (`twoHourlyDate` shifted by `twoHourlyDateOffsetMinutes`)
     */
    override fun isCoveringSubmissionDate(diagnosisKeySubmission: Instant, periodOffset: Duration): Boolean {
        val toExclusive = periodEndDateExclusive.plus(periodOffset)
        val fromInclusive = toExclusive.minus(Duration.ofHours(2))
        return (diagnosisKeySubmission.isAfter(fromInclusive) || diagnosisKeySubmission == fromInclusive)
            && diagnosisKeySubmission.isBefore(toExclusive)
    }

    /**
     * @return list of valid `TwoHourlyPeriod` ending with `this` `TwoHourlyPeriod`
     */
    override fun allPeriodsToGenerate(): List<TwoHourlyZIPSubmissionPeriod> {
        val periods = ArrayList<TwoHourlyZIPSubmissionPeriod>()
        var currentInstant = periodEndDateExclusive
        for (i in 0 until TOTAL_TWO_HOURLY_ZIPS) {
            periods.add(TwoHourlyZIPSubmissionPeriod(currentInstant))
            currentInstant = currentInstant.minus(Duration.ofHours(2))
        }
        return periods
    }

    override val startInclusive: Instant
        get() = periodEndDateExclusive.minus(Duration.ofHours(2))

    override val endExclusive: Instant
        get() = periodEndDateExclusive

    override fun toString(): String =
        "2 hours: from ${HOURLY_FORMAT.format(startInclusive)} (inclusive) to ${HOURLY_FORMAT.format(endExclusive)} (exclusive)"

    companion object {
        const val TWO_HOURLY_PATH_PREFIX = "distribution/two-hourly/"
        private const val TOTAL_TWO_HOURLY_ZIPS = ENIntervalNumber.MAX_DIAGNOSIS_KEY_AGE_DAYS * 12
        private val HOURLY_FORMAT = DateTimeFormatter.ofPattern("yyyyMMddHH").withZone(UTC)

        private fun requireValid(endDate: Instant): Instant {
            endDate.atZone(UTC).also {
                check(it[HOUR_OF_DAY] % 2 == 0)
                check(it[MINUTE_OF_HOUR] == 0)
                check(it[SECOND_OF_MINUTE] == 0)
                check(it[NANO_OF_SECOND] == 0)
            }
            return endDate
        }

        /**
         * @return end date (exclusive) of the two-hourly period comprising the Diagnosis Keys posted to the Submission Service at `diagnosisKeySubmission`
         */
        fun periodForSubmissionDate(diagnosisKeySubmission: Instant): TwoHourlyZIPSubmissionPeriod {
            val hour = diagnosisKeySubmission
                .truncatedTo(HOURS)
                .atZone(UTC)
                .hour
            return TwoHourlyZIPSubmissionPeriod(
                diagnosisKeySubmission
                    .truncatedTo(HOURS)
                    .atZone(UTC)
                    .withHour(hour - hour % 2)
                    .plus(Duration.ofHours(2))
                    .toInstant()
            )
        }
    }
}
